File: baoli_checkin.js
"use strict";

/*
Loon cron script.

Recommended argument examples:
token=YOUR_TOKEN&retry=3&timeout=20000&notify=1
mallId=3583&checkOnly=1&notify=0
storeKey=baoli_token&save=1&token=YOUR_TOKEN
mallStoreKey=baoli_mall_id
node=YourPolicyName
*/

const CONFIG = {
  baseUrl: "https://a.china-smartech.com",
  capturedToken: "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzdWIiOiI5NTExNTMxIiwiZ3VhcmQiOiJtZW1iZXIiLCJtYWxsX2lkIjoiMzU4MyIsImlzcyI6Imh0dHBzOi8vYS5jaGluYS1zbWFydGVjaC5jb20vcmVzdGZ1bC9tZW1iZXIveC90b2tlbiIsImV4cCI6NDA5MjU5OTM0OSwiaWF0IjoxNzczNTcwODQ1LCJhcHBfaWQiOiJ3eGI1ZTNmMzAwNzZmM2E2YzAifQ.qpxJv6hAJRflYTUL__tp3ro_4AXeTZ_XXGOl19QQRoE",
  defaultTimeoutMs: 20000,
  defaultRetry: 2,
  defaultNotify: true,
  defaultStoreKey: "baoli_token",
  defaultMallStoreKey: "baoli_mall_id",
  userAgent: "Mozilla/5.0 (iPhone; CPU iPhone OS 18_7 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15E148 MicroMessenger/8.0.69(0x18004531) NetType/4G Language/zh_CN",
  notificationOpenUrl: "loon://",
  successCodes: {
    0: true,
    200: true,
  },
};

function log(message) {
  if (typeof console !== "undefined" && console && typeof console.log === "function") {
    console.log("[baoli] " + message);
  }
  return message;
}

function sleep(ms) {
  return new Promise(function (resolve) {
    setTimeout(resolve, ms);
  });
}

function safeDecode(value) {
  try {
    return decodeURIComponent(String(value || "").replace(/\+/g, "%20"));
  } catch (_) {
    return String(value || "");
  }
}

function parseArgument(rawArgument) {
  const result = {};
  if (!rawArgument || typeof rawArgument !== "string") {
    return result;
  }

  rawArgument.split(/[&\n,]/).forEach(function (part) {
    const item = String(part || "").trim();
    if (!item) {
      return;
    }
    const index = item.indexOf("=");
    if (index === -1) {
      result[safeDecode(item)] = "1";
      return;
    }
    const key = safeDecode(item.slice(0, index).trim());
    const value = safeDecode(item.slice(index + 1).trim());
    if (key) {
      result[key] = value;
    }
  });

  return result;
}

function normalizeToken(token) {
  let value = String(token || "").replace(/[\r\n]/g, "").trim();
  if (!value) {
    return "";
  }
  if ((value.startsWith('"') && value.endsWith('"')) || (value.startsWith("'") && value.endsWith("'"))) {
    value = value.slice(1, -1).trim();
  }
  if (/^bearer\s+/i.test(value)) {
    value = value.replace(/^bearer\s+/i, "").trim();
  }
  return value;
}

function parseBoolean(value, defaultValue) {
  if (value === undefined || value === null || value === "") {
    return defaultValue;
  }
  const normalized = String(value).trim().toLowerCase();
  if (["1", "true", "yes", "on"].indexOf(normalized) !== -1) {
    return true;
  }
  if (["0", "false", "no", "off"].indexOf(normalized) !== -1) {
    return false;
  }
  return defaultValue;
}

function parseInteger(value, defaultValue, minValue) {
  const parsed = parseInt(value, 10);
  if (Number.isNaN(parsed)) {
    return defaultValue;
  }
  if (typeof minValue === "number" && parsed < minValue) {
    return defaultValue;
  }
  return parsed;
}

function normalizeTimeout(value, defaultValue) {
  const parsed = parseInteger(value, defaultValue, 1);
  if (parsed <= 120) {
    return parsed * 1000;
  }
  return parsed;
}

function safeInt(value, defaultValue) {
  const parsed = parseInt(value, 10);
  return Number.isNaN(parsed) ? defaultValue : parsed;
}

function hasPersistentStore() {
  return typeof $persistentStore !== "undefined" && $persistentStore && typeof $persistentStore.read === "function";
}

function readStore(key) {
  if (!hasPersistentStore() || !key) {
    return "";
  }
  return String($persistentStore.read(key) || "");
}

function writeStore(key, value) {
  if (!hasPersistentStore() || !key || !value || typeof $persistentStore.write !== "function") {
    return false;
  }
  try {
    return $persistentStore.write(String(value), String(key));
  } catch (_) {
    return false;
  }
}

function base64UrlToUtf8(value) {
  const base64 = String(value || "").replace(/-/g, "+").replace(/_/g, "/");
  const padded = base64 + "===".slice((base64.length + 3) % 4);

  if (typeof atob === "function") {
    const binary = atob(padded);
    let hex = "";
    for (let i = 0; i < binary.length; i += 1) {
      hex += "%" + ("00" + binary.charCodeAt(i).toString(16)).slice(-2);
    }
    return decodeURIComponent(hex);
  }

  const chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=";
  let output = "";
  let buffer = 0;
  let bits = 0;

  for (let i = 0; i < padded.length; i += 1) {
    const current = padded.charAt(i);
    if (current === "=") {
      break;
    }
    const index = chars.indexOf(current);
    if (index < 0) {
      continue;
    }
    buffer = (buffer << 6) | index;
    bits += 6;
    if (bits >= 8) {
      bits -= 8;
      output += String.fromCharCode((buffer >> bits) & 0xff);
    }
  }

  let hex = "";
  for (let i = 0; i < output.length; i += 1) {
    hex += "%" + ("00" + output.charCodeAt(i).toString(16)).slice(-2);
  }
  return decodeURIComponent(hex);
}

function decodeJwtPayload(token) {
  const parts = String(token || "").split(".");
  if (parts.length < 2) {
    throw new Error("token 不是有效的 JWT");
  }
  const payload = JSON.parse(base64UrlToUtf8(parts[1]));
  if (!payload || typeof payload !== "object") {
    throw new Error("token payload 解析失败");
  }
  return payload;
}

function isTokenExpired(payload) {
  const exp = safeInt(payload && payload.exp, 0);
  return exp > 0 && exp <= Math.floor(Date.now() / 1000);
}

function toText(data) {
  if (typeof data === "string") {
    return data;
  }
  if (data === undefined || data === null) {
    return "";
  }
  if (typeof data === "object" && typeof data.byteLength === "number") {
    const view = data instanceof Uint8Array ? data : new Uint8Array(data);
    let output = "";
    for (let i = 0; i < view.length; i += 1) {
      output += String.fromCharCode(view[i]);
    }
    try {
      return decodeURIComponent(
        output
          .split("")
          .map(function (char) {
            return "%" + ("00" + char.charCodeAt(0).toString(16)).slice(-2);
          })
          .join("")
      );
    } catch (_) {
      return output;
    }
  }
  return String(data);
}

function parseApiResult(bodyText, httpStatus) {
  let raw;
  try {
    raw = JSON.parse(bodyText);
  } catch (_) {
    throw new Error("接口未返回合法 JSON: " + bodyText);
  }

  if (!raw || typeof raw !== "object") {
    throw new Error("接口返回结构异常: " + bodyText);
  }

  return {
    httpStatus: safeInt(httpStatus, 0),
    code: safeInt(raw.code, -1),
    msg: String(raw.msg || ""),
    data: raw.data,
    raw: raw,
  };
}

function httpRequest(method, params) {
  return new Promise(function (resolve, reject) {
    if (typeof $httpClient === "undefined" || !$httpClient || typeof $httpClient[method] !== "function") {
      reject(new Error("当前环境不支持 Loon $httpClient"));
      return;
    }

    $httpClient[method](params, function (error, response, data) {
      if (error) {
        reject(new Error(String(error)));
        return;
      }
      resolve({
        status: response && response.status,
        headers: (response && response.headers) || {},
        data: data,
      });
    });
  });
}

async function requestJson(method, path, token, options) {
  const headers = {
    Accept: "application/json",
    Authorization: "Bearer " + token,
    "User-Agent": options.userAgent,
  };

  const request = {
    url: options.baseUrl + path,
    timeout: options.timeoutMs,
    headers: headers,
  };

  if (options.node) {
    request.node = options.node;
  }

  if (options.body !== undefined) {
    request.headers["Content-Type"] = "application/json";
    request.body = JSON.stringify(options.body);
  }

  const attempts = Math.max(1, options.retry + 1);
  let lastError = "";

  for (let attempt = 1; attempt <= attempts; attempt += 1) {
    try {
      const response = await httpRequest(method, request);
      return parseApiResult(toText(response.data), response.status);
    } catch (error) {
      lastError = error && error.message ? error.message : String(error);
      if (attempt < attempts) {
        await sleep(Math.min(2000, attempt * 1000));
      }
    }
  }

  throw new Error("请求失败，已重试 " + attempts + " 次: " + lastError);
}

function getTodayEntry(formData) {
  if (!formData || !Array.isArray(formData.days)) {
    return null;
  }
  for (let i = 0; i < formData.days.length; i += 1) {
    const item = formData.days[i];
    if (item && safeInt(item.today, 0) === 1) {
      return item;
    }
  }
  return null;
}

function formatFormSummary(formData) {
  const today = getTodayEntry(formData);
  if (!today) {
    return "未找到今日签到记录。";
  }

  const stat = formData && formData.stat ? formData.stat : {};
  const checked = safeInt(today.point_total, 0) > 0;
  const day = today.day || "今天";
  const pointTotal = safeInt(today.point_total, 0);
  const continuous = safeInt(stat.continuous_count, 0);
  return day + " " + (checked ? "已签到" : "未签到") + "，今日积分 " + pointTotal + "，连续签到 " + continuous + " 天。";
}

function isSuccess(result) {
  return Boolean(CONFIG.successCodes[result.code]);
}

function isAlreadyCheckedIn(result) {
  return String(result && result.msg || "").indexOf("已签到") !== -1;
}

function isAuthError(result) {
  const status = safeInt(result && result.httpStatus, 0);
  const code = safeInt(result && result.code, 0);
  return status === 401 || status === 403 || code === 401 || code === 403;
}

function getPoints(data) {
  if (!data || typeof data !== "object") {
    return {
      point: 0,
      totalPoint: 0,
    };
  }
  return {
    point: safeInt(data.point, 0),
    totalPoint: safeInt(data.total_point, 0),
  };
}

function postNotification(enabled, title, subtitle, content) {
  if (!enabled || typeof $notification === "undefined" || !$notification || typeof $notification.post !== "function") {
    return;
  }
  $notification.post(title, subtitle, content, CONFIG.notificationOpenUrl);
}

function buildRuntimeConfig() {
  const args = parseArgument(typeof $argument === "string" ? $argument : "");
  const storeKey = String(args.storeKey || args.store_key || CONFIG.defaultStoreKey || "").trim();
  const mallStoreKey = String(args.mallStoreKey || args.mall_store_key || CONFIG.defaultMallStoreKey || "").trim();

  if (parseBoolean(args.save || args.persist, false) && args.token) {
    const saved = writeStore(storeKey, normalizeToken(args.token));
    log(saved ? "已将 token 写入本地存储。" : "写入 token 到本地存储失败。");
  }

  if (parseBoolean(args.save || args.persist, false) && (args.mallId || args.mall_id)) {
    writeStore(mallStoreKey, String(args.mallId || args.mall_id).trim());
  }

  const token = normalizeToken(args.token || readStore(storeKey) || CONFIG.capturedToken);
  const notify = parseBoolean(args.notify, CONFIG.defaultNotify);
  const timeoutMs = normalizeTimeout(args.timeout, CONFIG.defaultTimeoutMs);
  const retry = parseInteger(args.retry, CONFIG.defaultRetry, 0);

  return {
    args: args,
    token: token,
    storeKey: storeKey,
    mallStoreKey: mallStoreKey,
    mallId: String(args.mallId || args.mall_id || readStore(mallStoreKey) || "").trim(),
    notify: notify,
    checkOnly: parseBoolean(args.checkOnly || args.check_only, false),
    timeoutMs: timeoutMs,
    retry: retry,
    node: String(args.node || "").trim(),
    baseUrl: CONFIG.baseUrl,
    userAgent: CONFIG.userAgent,
  };
}

async function runCheckIn(config) {
  if (!config.token) {
    throw new Error("缺少 token。请先启用抓 token 脚本并打开一次小程序签到页，或手动通过 argument 传 token=...");
  }

  let jwtPayload = null;
  try {
    jwtPayload = decodeJwtPayload(config.token);
  } catch (error) {
    if (!config.mallId) {
      throw new Error("token 解析失败，且没有提供 mallId。");
    }
    log("token 不是标准 JWT，改用 argument 里的 mallId。");
  }

  if (jwtPayload && isTokenExpired(jwtPayload)) {
    throw new Error("token 已过期，请重新抓包替换。");
  }

  const mallId = config.mallId || String((jwtPayload && jwtPayload.mall_id) || "").trim();
  if (!mallId) {
    throw new Error("无法确定 mallId。请先启用抓 token 脚本，或在 argument 中传 mallId=3583。");
  }

  let formError = "";

  try {
    const form = await requestJson("get", "/restful/mall/" + mallId + "/checkInForm?with_records=1", config.token, {
      baseUrl: config.baseUrl,
      timeoutMs: config.timeoutMs,
      retry: config.retry,
      userAgent: config.userAgent,
      node: config.node,
    });

    if (isSuccess(form) && form.data && typeof form.data === "object") {
      const summary = formatFormSummary(form.data);
      log(summary);
      const today = getTodayEntry(form.data);
      if (today && safeInt(today.point_total, 0) > 0) {
        const message = "今天已经签到，无需重复签到。";
        log(message);
        postNotification(config.notify, "保利签到", "无需重复签到", summary);
        return;
      }
    } else {
      formError = "查询签到状态失败: http=" + form.httpStatus + ", code=" + form.code + ", msg=" + form.msg;
    }
  } catch (error) {
    formError = error && error.message ? error.message : String(error);
  }

  if (config.checkOnly) {
    if (formError) {
      throw new Error(formError);
    }
    const message = "仅检查状态完成，今天还未签到。";
    log(message);
    postNotification(config.notify, "保利签到", "状态检查完成", message);
    return;
  }

  if (formError) {
    log(formError);
    log("查询失败，继续尝试直接签到。");
  }

  const sign = await requestJson("post", "/restful/mall/" + mallId + "/checkInRecord", config.token, {
    baseUrl: config.baseUrl,
    timeoutMs: config.timeoutMs,
    retry: config.retry,
    userAgent: config.userAgent,
    node: config.node,
    body: {
      latitude: 0,
      longitude: 0,
    },
  });

  if (isSuccess(sign)) {
    const points = getPoints(sign.data);
    const successMessage = "签到成功，获得 " + points.point + " 积分，累计 " + points.totalPoint + " 积分。";
    log(successMessage);
    postNotification(config.notify, "保利签到", "签到成功", successMessage);
    return;
  }

  if (isAlreadyCheckedIn(sign)) {
    const alreadyMessage = "今天已经签到。";
    log(alreadyMessage);
    postNotification(config.notify, "保利签到", "无需重复签到", alreadyMessage);
    return;
  }

  if (isAuthError(sign)) {
    throw new Error("签到失败：token 可能已失效，请重新抓包替换 token。");
  }

  throw new Error("签到失败: http=" + sign.httpStatus + ", code=" + sign.code + ", msg=" + sign.msg);
}

async function main() {
  const config = buildRuntimeConfig();
  await runCheckIn(config);
}

function isLoonRuntime() {
  return typeof $done === "function" && typeof $httpClient !== "undefined";
}

if (isLoonRuntime()) {
  main()
    .catch(function (error) {
      const message = error && error.message ? error.message : String(error);
      log(message);
      const args = parseArgument(typeof $argument === "string" ? $argument : "");
      const notify = parseBoolean(args.notify, CONFIG.defaultNotify);
      postNotification(notify, "保利签到", "执行失败", message);
    })
    .finally(function () {
      $done();
    });
} else {
  log("This script is intended to run inside Loon.");
}

File: baoli_checkin.py
#!/usr/bin/env python3
"""保利会员小程序自动签到脚本，仅依赖 Python 标准库。"""

from __future__ import annotations

import argparse
import base64
import json
import os
import socket
import sys
import time
import urllib.error
import urllib.request
from dataclasses import dataclass
from typing import Any


BASE_URL = "https://a.china-smartech.com"
# 直接替换成最新抓包里的 token，不要带 Bearer 前缀。
CAPTURED_TOKEN = "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJzdWIiOiI5NTExNTMxIiwiZ3VhcmQiOiJtZW1iZXIiLCJtYWxsX2lkIjoiMzU4MyIsImlzcyI6Imh0dHBzOi8vYS5jaGluYS1zbWFydGVjaC5jb20vcmVzdGZ1bC9tZW1iZXIveC90b2tlbiIsImV4cCI6NDA5MjU5OTM0OSwiaWF0IjoxNzczNTcwODQ1LCJhcHBfaWQiOiJ3eGI1ZTNmMzAwNzZmM2E2YzAifQ.qpxJv6hAJRflYTUL__tp3ro_4AXeTZ_XXGOl19QQRoE"
DEFAULT_USER_AGENT = (
    "Mozilla/5.0 (iPhone; CPU iPhone OS 18_7 like Mac OS X) "
    "AppleWebKit/605.1.15 (KHTML, like Gecko) Mobile/15E148 "
    "MicroMessenger/8.0.69(0x18004531) NetType/4G Language/zh_CN"
)
DEFAULT_TIMEOUT = 15
DEFAULT_RETRY = 2
SUCCESS_CODES = {0, 200}


@dataclass
class ApiResult:
    http_status: int
    code: int
    msg: str
    data: Any
    raw: dict[str, Any]


def log(message: str, *, error: bool = False) -> None:
    print(message, file=sys.stderr if error else sys.stdout)


def normalize_token(token: str) -> str:
    token = (token or "").replace("\r", "").replace("\n", "").strip()
    if not token:
        return ""
    if (token.startswith('"') and token.endswith('"')) or (token.startswith("'") and token.endswith("'")):
        token = token[1:-1].strip()
    if token.lower().startswith("bearer "):
        token = token[7:].strip()
    return token


def parse_positive_int(value: str | None, default: int) -> int:
    if value is None or value == "":
        return default
    try:
        parsed = int(value)
    except (TypeError, ValueError):
        return default
    return parsed if parsed > 0 else default


def parse_non_negative_int(value: str | None, default: int) -> int:
    if value is None or value == "":
        return default
    try:
        parsed = int(value)
    except (TypeError, ValueError):
        return default
    return parsed if parsed >= 0 else default


def safe_int(value: Any, default: int = 0) -> int:
    try:
        return int(value)
    except (TypeError, ValueError):
        return default


def decode_jwt_payload(token: str) -> dict[str, Any]:
    try:
        _, payload, _ = token.split(".", 2)
        payload += "=" * (-len(payload) % 4)
        decoded = json.loads(base64.urlsafe_b64decode(payload))
        if not isinstance(decoded, dict):
            raise ValueError
        return decoded
    except Exception as exc:  # noqa: BLE001
        raise ValueError("token 不是有效的 JWT，无法解析 mall_id") from exc


def is_token_expired(payload: dict[str, Any]) -> bool:
    exp = safe_int(payload.get("exp"), 0)
    return exp > 0 and exp <= int(time.time())


def decode_response_body(response: Any, raw_body: bytes) -> str:
    charset = "utf-8"
    headers = getattr(response, "headers", None)
    if headers is not None:
        charset = headers.get_content_charset() or charset
    return raw_body.decode(charset, errors="replace")


def parse_api_result(content: str, http_status: int) -> ApiResult:
    try:
        raw = json.loads(content)
    except json.JSONDecodeError as exc:
        raise RuntimeError(f"接口未返回合法 JSON: {content}") from exc

    if not isinstance(raw, dict):
        raise RuntimeError(f"接口返回结构异常: {content}")

    return ApiResult(
        http_status=http_status,
        code=safe_int(raw.get("code"), -1),
        msg=str(raw.get("msg", "")),
        data=raw.get("data"),
        raw=raw,
    )


def request_json(
    method: str,
    path: str,
    token: str,
    *,
    payload: dict[str, Any] | None = None,
    timeout: int = DEFAULT_TIMEOUT,
    retry: int = DEFAULT_RETRY,
    user_agent: str = DEFAULT_USER_AGENT,
) -> ApiResult:
    body = None
    headers = {
        "Accept": "application/json",
        "Authorization": f"Bearer {token}",
        "User-Agent": user_agent,
    }

    if payload is not None:
        body = json.dumps(payload, separators=(",", ":")).encode("utf-8")
        headers["Content-Type"] = "application/json"

    request = urllib.request.Request(
        url=f"{BASE_URL}{path}",
        data=body,
        headers=headers,
        method=method,
    )

    attempts = max(1, retry + 1)
    last_error = ""
    for attempt in range(1, attempts + 1):
        try:
            with urllib.request.urlopen(request, timeout=timeout) as response:
                content = decode_response_body(response, response.read())
                return parse_api_result(content, response.getcode())
        except urllib.error.HTTPError as exc:
            content = decode_response_body(exc, exc.read())
            return parse_api_result(content, exc.code)
        except (urllib.error.URLError, socket.timeout, TimeoutError) as exc:
            reason = getattr(exc, "reason", exc)
            last_error = str(reason)
        except Exception as exc:  # noqa: BLE001
            last_error = str(exc)

        if attempt < attempts:
            time.sleep(min(2, attempt))

    raise RuntimeError(f"请求失败，已重试 {attempts} 次: {last_error or '未知错误'}")


def get_today_entry(form_data: dict[str, Any]) -> dict[str, Any] | None:
    for item in form_data.get("days", []):
        if item.get("today") == 1:
            return item
    return None


def format_form_summary(form_data: dict[str, Any]) -> str:
    today = get_today_entry(form_data)
    stat = form_data.get("stat") or {}
    if not today:
        return "未找到今日签到记录。"

    checked = safe_int(today.get("point_total"), 0) > 0
    today_str = today.get("day", "今天")
    continuous = safe_int(stat.get("continuous_count"), 0)
    point_total = safe_int(today.get("point_total"), 0)
    status = "已签到" if checked else "未签到"
    return f"{today_str} {status}，今日积分 {point_total}，连续签到 {continuous} 天。"


def is_success(result: ApiResult) -> bool:
    return result.code in SUCCESS_CODES


def is_already_checked_in(result: ApiResult) -> bool:
    return "已签到" in result.msg


def is_auth_error(result: ApiResult) -> bool:
    auth_codes = {401, 403}
    return result.http_status in auth_codes or result.code in auth_codes


def get_points(data: Any) -> tuple[int, int]:
    if not isinstance(data, dict):
        return 0, 0
    return safe_int(data.get("point"), 0), safe_int(data.get("total_point"), 0)


def check_in(
    token: str,
    mall_id: str,
    *,
    timeout: int,
    retry: int,
    check_only: bool,
    json_output: bool,
) -> int:
    form: ApiResult | None = None
    form_error = ""

    try:
        form = request_json(
            "GET",
            f"/restful/mall/{mall_id}/checkInForm?with_records=1",
            token,
            timeout=timeout,
            retry=retry,
        )
    except RuntimeError as exc:
        form_error = str(exc)
    else:
        if is_success(form) and isinstance(form.data, dict):
            if json_output:
                log(json.dumps({"stage": "form", **form.raw}, ensure_ascii=False, indent=2))
            else:
                log(format_form_summary(form.data))

            today = get_today_entry(form.data)
            if today and safe_int(today.get("point_total"), 0) > 0:
                log("今天已经签到，无需重复签到。")
                return 0
        else:
            form_error = f"查询签到状态失败: http={form.http_status}, code={form.code}, msg={form.msg}"

    if form_error:
        log(form_error, error=True)
        if check_only:
            return 1
        log("查询失败，继续尝试直接签到。")

    sign = request_json(
        "POST",
        f"/restful/mall/{mall_id}/checkInRecord",
        token,
        payload={"latitude": 0, "longitude": 0},
        timeout=timeout,
        retry=retry,
    )

    if json_output:
        log(json.dumps({"stage": "checkin", **sign.raw}, ensure_ascii=False, indent=2))

    if is_success(sign):
        point, total_point = get_points(sign.data)
        if not json_output:
            log(f"签到成功，获得 {point} 积分，累计 {total_point} 积分。")
        return 0

    if is_already_checked_in(sign):
        if not json_output:
            log("今天已经签到。")
        return 0

    if is_auth_error(sign):
        log("签到失败：token 可能已失效，请重新抓包替换 token。", error=True)
        if json_output:
            log(json.dumps(sign.raw, ensure_ascii=False, indent=2), error=True)
        return 1

    log(f"签到失败: http={sign.http_status}, code={sign.code}, msg={sign.msg}", error=True)
    if json_output:
        log(json.dumps(sign.raw, ensure_ascii=False, indent=2), error=True)
    return 1


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description="保利会员小程序自动签到")
    parser.add_argument(
        "--token",
        help="Bearer token，不带 Bearer 前缀。默认读取 BAOLI_TOKEN 环境变量，再回退到当前抓包里的 token。",
    )
    parser.add_argument("--mall-id", help="商场 ID；默认从 JWT 的 mall_id 自动解析。")
    parser.add_argument("--timeout", type=int, default=DEFAULT_TIMEOUT, help="请求超时时间，默认 15 秒。")
    parser.add_argument("--retry", type=int, default=DEFAULT_RETRY, help="网络失败后的额外重试次数，默认 2 次。")
    parser.add_argument("--check-only", action="store_true", help="只查询今日是否已签到，不发起签到。")
    parser.add_argument("--json", action="store_true", help="输出原始 JSON 结果。")
    return parser


def main() -> int:
    args = build_parser().parse_args()

    token = normalize_token(args.token or os.environ.get("BAOLI_TOKEN") or CAPTURED_TOKEN)
    if not token:
        log("缺少 token，请通过 --token 或 BAOLI_TOKEN 提供。", error=True)
        return 1

    try:
        payload = decode_jwt_payload(token)
    except ValueError as exc:
        log(str(exc), error=True)
        return 1

    if is_token_expired(payload):
        log("token 已过期，请重新抓包替换。", error=True)
        return 1

    mall_id = args.mall_id or str(payload.get("mall_id") or "").strip()
    if not mall_id:
        log("无法从 token 中解析 mall_id，请通过 --mall-id 指定。", error=True)
        return 1

    timeout = parse_positive_int(os.environ.get("BAOLI_TIMEOUT"), args.timeout)
    retry = parse_non_negative_int(os.environ.get("BAOLI_RETRY"), args.retry)

    try:
        return check_in(
            token=token,
            mall_id=mall_id,
            timeout=timeout,
            retry=retry,
            check_only=args.check_only,
            json_output=args.json,
        )
    except RuntimeError as exc:
        log(str(exc), error=True)
        return 1
    except KeyboardInterrupt:
        log("已手动中断。", error=True)
        return 130
    except Exception as exc:  # noqa: BLE001
        log(f"发生未预期错误: {exc}", error=True)
        return 1


if __name__ == "__main__":
    raise SystemExit(main())
